File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
import java.util.*;

class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        long answer = 0;
        int deliveryIdx = n - 1;
        int pickupIdx = n - 1;
        while(deliveryIdx != -1 || pickupIdx != -1){
            int tmp = cap;
            boolean isFirst = true;
            int dist = 0;
            int i = 0;
            for(i = deliveryIdx; i >= 0; i--){
                if(deliveries[i] == 0) continue;
                if(isFirst){
                    dist = i+1;
                    isFirst = false;
                }
                if(tmp - deliveries[i] <= 0){
                    deliveries[i] -= tmp;
                    break;
                }
                tmp -= deliveries[i];
                deliveries[i] = 0;
            }
            deliveryIdx = i;
            isFirst = true;
            tmp = cap;
            
            for(i = pickupIdx; i >= 0; i--){
                if(pickups[i] == 0) continue;
                if(isFirst){
                    dist = Math.max(i+1, dist);
                    isFirst = false;
                }
                if(tmp - pickups[i] <= 0){
                    pickups[i] -= tmp;
                    break;
                }
                tmp -= pickups[i];
                pickups[i] = 0;
            }
            pickupIdx = i;
            answer += (long)(2*dist);
        }
        return answer;
    }
}
// 그리디
// 시간 복잡도: n^2
// deliveryCursor = n - 1
// pickupCursor = n - 1

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.py
def solution(cap, n, deliveries, pickups):
    answer = 0
    delivery = 0 # delivery < 0: 배달 공간 여유가 있는거
    pickup = 0 # pickup < 0: 수거 공간 여유가 있는거
    for i in range(n-1, -1, -1):
        delivery += deliveries[i]
        pickup += pickups[i]
        
        while delivery > 0 or pickup > 0:
            delivery -= cap
            pickup -= cap
            
            answer += (i+1) * 2
    
    return answer

# 그리디
